### DISCUSSION TITLE
Prevent destruction/re-creation of tab content in TabControl when switching tabs

### DISCUSSION THREAD

**Discussion by StefanKoell** [AvaloniaUI/Avalonia#11837] [Q&A] [answered]
Hi!

Is there a way to tell the TabControl to not destroy/re-create the content of a tab item when switching between the tabs?

Thanks,
Stefan

**Comment by timunie** [comment]
Normally, in an MVVM App, should not be a big deal. However, I think in 11.0 this should be handled better (iirc).

**Reply by StefanKoell** [reply]
I'm actually using 11.0 (RC) already. Unfortunately I can't really see a way to prevent that. Maybe I'm missing something but in my case, where I instantiate RDP connections using a NativeControlHost, it's not really desirable to destroy the childs in the visual tree. Any idea how I can prevent that from happening?

**Comment by maxkatz6** [comment]
No, you can't change this behavior for the TabControl. 
You can use a combination of TabStrip control with headers and ContentControl to host content, and implement simple content management this way, for example.

**Reply by StefanKoell** [reply]
Hi @maxkatz6! Thanks for chiming in. Can you expand on "implement simple content management" with a simple example? Not sure I understand what you mean.

**Reply by timunie** [reply]
For example you can just set IsVisible=[True/False] on several controls stacked above each other?

**Reply by StefanKoell** [reply]
Please forgive my ignorance. I'm not sure how I can make this work. I basically have a VM with "tabs" of different types. Each type  instantiates a UserControl containing a NativeControlHost descendent. So I have a dynamic number of different tabs which I need to preserve while switching between them. The main problem I'm facing is that I don't have control over whether or not my NativeControlHost descendent is disposed/destroyed.

**Reply by timunie** [reply]
Sorry, I don't know your exact usecase. I assumed you know the items in advance. Probably you can listen to `AttachedToVisualTree` in order to add the views into some kind of cache logic? Just an idea, I've never faced this issue before.

**Reply by StefanKoell** [reply]
I tried hooking into OnDetachedFromVisualTree but the main issue is that I have no means to control the lifetime of my native (WinForms) control. Looking at the source code, there's no way to prevent the NativeControlHost to call DestroyNativeControl and handle the lifetime myself.

I find it odd that it behaves this way because native controls (such as WinForms controls) do not have similar data binding and do not really benefit or can make use of the mechanism XAML has with the efficient logical tree to visual tree management.

**Reply by timunie** [reply]
well, if you have a good API proposal here and are willing to provide the needed PRs, I think we are open for it. @maxkatz6 may disagree though.

**Reply by StefanKoell** [reply]
I'm happy to participate. Since I'm fairly new to Avalonia I just wanted to make sure I understand certain design decisions and maybe there's a solution to my problem I'm just not seeing. I will try to extend/change the NativeControlHost and see if I can make this work.

**Reply by maxkatz6** [reply]
>  find it odd that it behaves this way because native controls

It's not about native framework (winforms...), but about native views. WinForm interop creates an embeddable window, which needs to be disposed/destroyed, when it's not used. And there is no better way to trigger it than control's detached/unloaded events.

> You can use a combination of TabStrip control with headers and ContentControl to host content, and implement simple content management this way, for example.

TabStrip behaves in the same way as TabControl headers, so you have tabs and selected tab information from it.
ContentControl can be used to host tab content, but it only has a single ContentControl.Content property. As an option, you can handle TabStrip selection changed events and create+cache native control host (tab contents) and assign it to the ContentControl.Content.

**Reply by StefanKoell** [reply]
Well, I just tried to create my own NativeControlHost based on the shipped but it seems that there are properties and methods of event args and public interfaces marked internal. Not sure why. So even though the NativeControlHost is only 180 lines of code, I can't really roll my own without cloning/building the whole repository. Bummer...

**Reply by maxkatz6** [reply]
Why do you need to create your own NativeControlHost?

**Reply by StefanKoell** [reply]
> It's not about native framework (winforms...), but about native views. WinForm interop creates an embeddable window, which needs to be disposed/destroyed, when it's not used. And there is no better way to trigger it than control's detached/unloaded events.

Sure, in many cases this is ok but there are use cases, especially when hosting the native views in tabs, there's probably a need to keep those instances alive.

To answer your question why I want to create my own NativeControlHost:
The idea was simply to get more control over the lifetime of my instance. Right now I see no way to solve my issue. Sure I need to take care of the disposal and cleanup myself but that's OK.

**Reply by StefanKoell** [reply]
Found this for WPF:
https://ikriv.com/dev/wpf/PersistentTabControl/

 Maybe I'm able to adapt it for Avalonia...

**Reply by StefanKoell** [reply]
No cigar :( I guess I have to give up.

First I was pretty excited when I got my WinForms controls to work with Avalonia and I was sure I can ditch WPF once and for all but this is a show stopper for me unfortunately. I don't know enough of Avalonia to resolve the issue and it seems that the framework at this stage cannot fulfill my requirements. I feel like it's so close and just a small step away. Hopefully someone can help me find a solution.

**Reply by timunie** [reply]
Your chance increases if you have a minimal sample to reproduce the issue.

**Reply by StefanKoell** [reply]
Sure, no problem. I've created a stripped down version of my test app.

There's a Connect->Test menu to create tabs.
Simply create two tabs by invoking the menu two times.
Each tab will get a random color and instance number.
There's also a text box which clearly illustrates the lost of state when switching tabs.

So put something in the textbox on tab 2 and switch to tab 1 and back to tab 2. You will notice a different color, instance number and the entered text is gone. Because everytime a tab is switched the previous instance gets destroyed and a new instance is created.

This behavior is problematic for interop scenario like mine.

Thanks for looking into it. I really appreciate it!

[AvaloniaInterop.zip](https://github.com/AvaloniaUI/Avalonia/files/11798549/AvaloniaInterop.zip)

**Reply by timunie** [reply]
Sorry, I'm unable to open any new tab. If I've more time, I'll look into it the other day. Maybe also some one else will want to take a look.

**Reply by StefanKoell** [reply]
That's weird. I just downloaded the zip again and tried it on my machine. I do use JetBrains rider though and since I'm on an ARM64 machine, the RuntimeIdentifier should be changed for x64 if needed.
Do you get an error?

**Reply by timunie** [reply]
`System.InvalidOperationException: Unable to create child window for native control host. Application manifest with supported OS list might be required.`

**Reply by StefanKoell** [reply]
I'm assuming you are on x64, right? Can you try to change the RID to:

` <RuntimeIdentifier>win10-x64</RuntimeIdentifier>`

in the Test.csproj?

**Reply by timunie** [reply]
looks like a manifest file is missing.

**Reply by timunie** [reply]
finally

**Comment by timunie** [comment, marked as answer]
Okay looks like the idea with IsVisible works for me. Here is what I did:

in MainView:

```xml 
    <Grid RowDefinitions="Auto,Auto,*" >

        <Menu Grid.Row="0">
            <MenuItem Header="_Action">
                <MenuItem Header="_Connect">
                    <MenuItem Header="Test" Command="{Binding AddTest}" />
                </MenuItem>
                <MenuItem Header="_Disconnect" Command="{Binding Close}" />
            </MenuItem>
        </Menu>

        <TabStrip Name="TabControl" ItemsSource="{Binding Tabs}" SelectedItem="{Binding SelectedTab}" Grid.Row="1" >
            <TabStrip.ItemTemplate>
                <DataTemplate>
                    <TextBlock Text="xx{Binding Caption}"></TextBlock>
                </DataTemplate>
            </TabStrip.ItemTemplate>
        </TabStrip>
        
        <ItemsControl Grid.Row="2" ItemsSource="{Binding Tabs}" >
            <ItemsControl.ItemsPanel>
                <ItemsPanelTemplate>
                    <Panel></Panel>
                </ItemsPanelTemplate>
            </ItemsControl.ItemsPanel>
            <ItemsControl.ItemTemplate>
                <views1:ViewLocator />
            </ItemsControl.ItemTemplate>
        </ItemsControl>
    </Grid>
```

In IsEqualConverter:
```cs
public class IsEqualConverter : IMultiValueConverter
{
    public static IsEqualConverter Instance { get; } = new();

    public object? Convert(IList<object?> values, Type targetType, object? parameter, CultureInfo culture)
    {
        return values?.Count == 2 && values[0] == values[1];
    }
}
```

in TestView:

```xml
    <UserControl.IsVisible>
        <MultiBinding Converter="{x:Static viewModels:IsEqualConverter.Instance}">
            <Binding ></Binding>
            <Binding RelativeSource="{RelativeSource Mode=FindAncestor, AncestorType=ItemsControl}"
                     Path="DataContext.SelectedTab"></Binding>
        </MultiBinding>
    </UserControl.IsVisible>
    
    <views:TestNative Name="TestNative" />
```

works for me. All controls are being kept in cache.

**Reply by StefanKoell** [reply]
First of all, thanks so much for taking a look and helping out!

Any chance you can upload the modified project? I implemented your stuff but it doesn't seem to work for me. I noticed
`                <views1:ViewLocator />`

which is different from my project. Did you create a different ViewLocator?

**Reply by StefanKoell** [reply]
Sorry, my bad. Found the culprit. Got it working. Thanks again! This has been very helpful and also thanks for the patience!

cheers,
Stefan

**Comment by StefanKoell** [comment]
I also found a solution by hacking around the NativeControlHost and keeping instances of the native controls cached (see attachment with updated project).

I'm wondering what you guys think is the better approach?
Any comments or suggestions?

I can see the NativeControlHost hack advantageous because it would work "everywhere" (even the TabControl).
Whereas the ItemsControl approach allows me to implement the NativeControlHost as it's supposed to be and handles lifetime directly.

In any case, I now have two ways to deal with that so in general I guess my issue is resolved.

[AvaloniaInterop.zip](https://github.com/AvaloniaUI/Avalonia/files/11801117/AvaloniaInterop.zip)

**Reply by StefanKoell** [reply]
Just wanted to let people know that I created a nuget package with a custom NativeControlHost to handle lifetime of WinForms controls:
https://github.com/royalapplications/royalapps-community-avalonia

My preference is still the solution provided by @timunie using the ItemsControl through visibility switching. Only downside of this approach is that you are constraint to the ItemsControl which might be a problem if you are using 3rd party controls where you can't really change the way it behaves.

**Reply by timunie** [reply]
@StefanKoell in Avalonia you most of the time can override the Controltemplate. So you have a huge degree of freedom. For example, you can tweak TabControl to have an ItemsControl inside. I just was too lazy to do. 

Thanks for the 💐 anyway

**Reply by StefanKoell** [reply]
@timunie that's interesting. I was wondering if that's possible. I've seen some examples and it's worth looking into it. I'm still learning Avalonia/XAML and right now I'm also still in the prototyping phase but when it's time to work on that, I will definitely look into it. 

Coming from WinForms, interop with that technology is still important for me - at least at the beginning. I have some dependencies I can't get rid of quickly. I'm relieved that there are now two solutions to work with those dependencies and maybe someone finds this thread helpful.
